### ISSUE TITLE
Class hash does not exist

### ISSUE THREAD

**Issue by dragonfire1119** [lorisleiva/laravel-actions#4]
I'm getting this ``Class hash does not exist`` when running composer update.

Laravel 5.8: Failing
Laravel 5.7: Successful

Found out it was this package with:

```
"extra": {
    "laravel": {
      "dont-discover": [
        "lorisleiva/laravel-actions"
      ]
    }
  },
```

Specs:
Laravel Framework 5.8.19

**Comment by lorisleiva** [comment]
Hi there 👋 

Thanks for reporting this. I am not sure I understand the issue though. Are you explicitly removing auto-discovery of the package? If so, are you adding the service provider manually in your `config/app.php` file?

**Comment by dragonfire1119** [comment]
Yes, I'm removing the package from auto discovery temporary to find out which package is the problem.

When I comment out https://github.com/lorisleiva/laravel-actions/blob/9e44cf087400f73c2ddc4cf148ecf89f5f7be7cb/src/ActionServiceProvider.php#L15

It starts working again.

**Comment by lorisleiva** [comment]
Argh, that's annoying. I haven't manually tested it with a 5.8 application yet so I'm going to try with a fresh 5.8 Laravel install and see if I can reproduce the issue. Thanks again for reporting this bug, I'll keep you posted.

**Comment by lorisleiva** [comment]
Okay so it works with a fresh 5.8 install but I think the reason you're getting this is that some resources (the Hash class in this case) have not yet been registered in the container. 

So I'm pretty sure that moving the `events` singleton to the `boot` method will make it work (if you could double check that that'd be great 🙏).

However, that's not really a long term solution as we want to ensure the event dispatcher is ready for every boot method of every service provider. Does that make any sense?

Could you maybe share some information on the order of your service providers?

**Comment by lorisleiva** [comment]
After some deep digging into the Laravel framework, I've found the exact order that Laravel uses to load providers ([View source](https://github.com/laravel/framework/blob/0a12baf129c525d1a3f118c94f93a3bc97190779/src/Illuminate/Foundation/Application.php#L574-L585)):

```php
public function registerConfiguredProviders()
{
    $providers = Collection::make($this->config['app.providers'])
                    ->partition(function ($provider) {
                        return Str::startsWith($provider, 'Illuminate\\');
                    });
    $providers->splice(1, 0, [$this->make(PackageManifest::class)->providers()]);
    (new ProviderRepository($this, new Filesystem, $this->getCachedServicesPath()))
                ->load($providers->collapse()->toArray());
}
```

Basically, we have the following order:
1. Base (low-level) service providers (registered when the application singleton is being created).
2. Service providers from the `config/app.php` file that starts with `Illuminate\`.
3. Service providers auto-discovered from packages (including Laravel Action ⚡️).
4. The other service providers from the `config/app.php` file.

Therefore we should be fine overriding the event dispatcher in the `register` method instead of the `boot` method.

Do you have any complex-ish custom service providers in your application?

**Comment by dragonfire1119** [comment]
Hello, @lorisleiva thanks for looking into this! The only complex service provider I have is one & I register bind & singleton in it.

I moved that block of code down to the boot method & it works now.

I temporarily took out that custom service provider & it didn't change anything though.

**Comment by huy-tran** [comment]
I am having the same issue.

At the moment I moved the event singleton to boot method as @lorisleiva mentioned above. I have so many other Service Providers so I don't think I can take them out.

I love this package so hopefully we have a better solution soon.

Cheers!

**Comment by lorisleiva** [comment]
Hey, thanks for the additional information. 👍 

I'm definitely leaning toward **moving the event singleton to the `boot` method**.

My only concern is that it might cause other issues if another service provider has a boot method that depends on it and it hasn't been booted yet.

I've had a look at other packages and I'd say their bindings are 50/50 defined in `register`/`boot`. I'm just a bit more cautious here as this is an override of a core Laravel class.

Just a few questions before updating the source code:

- Can anyone foresee some issues with moving the event singleton to the boot method?
- Does it make sense to move all bindings to the `boot` method for consistency?
- Can anyone think of a better way to extend this event dispatcher altogether?

Thanks again for your help! 🌟

**Comment by deviouspk** [comment]
I don't really like the idea to use actions as listeners | jobs. It clearly violates the single responsability principle. Besides that the implementation is also horrible :( overriding the laravel dispatchers is not something you want to do (especially not from a third party package).

If you do want to go this way. You might want to consider disabling this behavior by default. And allow people to register it themselves manually in the boot() method.

**Comment by lorisleiva** [comment]
@deviouspk Nah, that's the whole point of Laravel Actions. I'm okay with an alternative solution but I'm not compromising on the current feature set.

**Comment by deviouspk** [comment]
Well i don't think there's another way to do this besides making your own dispatcher class.

Well then i would still not override the dispatchers. you can make your own listener: ActionEventListener that contains a static class $events.

You could Listen to events like this MyAction::listen(SomeEvent::class);
then the listen method would put it on the static event attribute and subscribe to that event.

![image](https://user-images.githubusercontent.com/7100315/58707376-06ec9d00-83b5-11e9-86b2-dc592150e093.png)


You can do more or less the same thing for jobs.

**Comment by lorisleiva** [comment]
I'm afraid this would compromise the current feature set by removing the ability to register actions as proper listeners. This is not the purpose of this thread.

**Comment by deviouspk** [comment]
what benefit do you get from running the action class itself as a listener?

**Comment by lorisleiva** [comment]
Fully leverage the Laravel API that we all know. For example:

```php
class EventServiceProvider extends ServiceProvider
{
    protected $listen = [
        'App\Events\MyEvent' => [
            'App\Action\MyAction',
        ],
    ];
}
```

**Comment by deviouspk** [comment]
yeah there's no way around that. The dispatcher handles that logic :s

**Comment by dragonfire1119** [comment]
@lorisleiva There must have been something that changed in 5.8 from 5.7 to make this error out? Because the exact same project with the exact same providers are working on 5.7 but not 5.8.

**Comment by lorisleiva** [comment]
Thanks @dragonfire1119. Unfortunately, I looked through the GitHub compare between 5.7 and 5.8 and I could not find any significant changes in either the event dispatcher or the way Laravel registers providers.

However, that little digging gave me some ideas to solve this issue.

I'd like to start by using the `extends` method instead of the `singleton` method, even though we just provide a new instance of the class (just like we currently override the bus dispatcher).

```php
// Before
$this->app->singleton('events', function ($app) {/* ... */});

// After
$this->app->extends('events', function ($dispatcher, $app) {/* ... */});
```

My only issue with that is I cannot reproduce the `Class hash does not exist` error. Could you please help me to reproduce this error (from a fresh Laravel 5.8 install)? If this turns out to be tricky, could you please try the solution mentioned above and let me know if it works?

We'll get to the bottom of this! 🤞

**Comment by lorisleiva** [comment]
After having read the implementation of the `extends()` method from the container, I feel pretty confident about this implementation.

I have therefore made the changes to the master branch. Could anyone that has this error confirm that it fixes the issue using `dev-master`?

**Comment by syropian** [comment]
Unfortunately I'm getting the same error using `composer require lorisleiva/laravel-actions:dev-master`. My app isn't a completely fresh 5.8 install, but it's pretty new.

**Comment by lorisleiva** [comment]
Thank you for letting me know @syropian. Let the digging continue haha.

**Comment by syropian** [comment]
@lorisleiva Do you want to try debugging the issue on my project? It's just a side project, and like I said, it's relatively fresh.

**Comment by lorisleiva** [comment]
@syropian Oh yes please! 🙏

**Comment by syropian** [comment]
@lorisleiva Here you go! https://github.com/syropian/under-wraps-server

Let me know if you have any questions around the setup. It's pretty standard afaik.

**Comment by lorisleiva** [comment]
@syropian Awesome, thanks. I'll do a fork to make sure I don't break anything in your app. 🙂

**Comment by dragonfire1119** [comment]
@lorisleiva I tried the dev-master of my product & it still throws ``Class hash does not exist``.

**Comment by lorisleiva** [comment]
Okay, so here is what is happening.

Starting from 5.8, Laravel adds a `rebinding` listeners on the `events` alias.

```php
// Illuminate\Auth\AuthServiceProvider
$this->app->rebinding('events', function ($app, $dispatcher) {
    if (! $app->resolved('auth')) {
        return;
    }
    if (method_exists($guard = $app['auth']->guard(), 'setDispatcher')) {
        $guard->setDispatcher($dispatcher);
    }
});
```

When we rebind the `events` instance from the container (whether it's via `singleton` or `extend`), it will run a `rebound` method that calls all listeners attached to `events`.

This will thus call the closure above and try to resolve `$app['auth']->guard()` which further down the line creates an `EloquentUserProvider` that requires an instance of `$this->app['hash']`. Phew.

The reason why this is a problem is that, at this stage, in container land, binding implementations are stored as closures until they are resolved. So wakening that rebound listener causes resolving some instances too early.

**Conclusion:**
Since rebinding listeners assume the container is "ready", we have to rebound once the container is ready. This means moving the `extend` methods to the `boot` method.

Btw, if you're wondering why the `extend` method for the bus dispatcher is not causing any problem, it's because its original service provider is deferred.

**TL;DR**
Moving everything from the `boot` method to the `register` method is the only way around this issue.

**Comment by dragonfire1119** [comment]
@lorisleiva Solved my issue right away moving events to boot method. Was just starting to look at what's different in 5.8 because I could not believe that 5.7 works & 5.8 does not work with no changes to internals. Glad you figured it out!

**Comment by lorisleiva** [comment]
New release [`v0.1.2`](https://github.com/lorisleiva/laravel-actions/releases/tag/v0.1.2) tagged.

Thanks for your help everyone. 🍺 

Feel free to reopen this issue if the new release causes side effects.
